File: vehicle-feature-catalog/src/python/vehicle_feature_catalog/facade.py
"""Public facade for the vehicle feature catalog.

This is the only file external code should import from (via the package
`__init__.py` re-export).
"""

from __future__ import annotations

from pathlib import Path
from typing import Iterable

from .entities import Feature, Make, Model, Trim, ValidationResult
from .errors import NotFoundError
from .indexes import CatalogIndexes
from .loader import YAMLLoader
from .validator import Validator


class VehicleFeatureCatalog:
    """Read-only in-memory catalog of Makes, Models, Trims, Features, and the matrix."""

    def __init__(self, indexes: CatalogIndexes) -> None:
        self._idx = indexes

    @classmethod
    def load(cls, data_dir: str | Path) -> "VehicleFeatureCatalog":
        path = Path(data_dir)
        raw = YAMLLoader.read_all(path)
        idx = CatalogIndexes.build(
            makes=raw.makes,
            models=raw.models,
            trims=raw.trims,
            features=raw.features,
            trim_features=raw.trim_features,
        )
        return cls(idx)

    def get_make(self, make_id: str) -> Make:
        try:
            return self._idx.makes_by_id[make_id]
        except KeyError:
            raise NotFoundError(f"Make not found: {make_id!r}") from None

    def get_model(self, model_id: str) -> Model:
        try:
            return self._idx.models_by_id[model_id]
        except KeyError:
            raise NotFoundError(f"Model not found: {model_id!r}") from None

    def get_trim(self, trim_id: str) -> Trim:
        try:
            return self._idx.trims_by_id[trim_id]
        except KeyError:
            raise NotFoundError(f"Trim not found: {trim_id!r}") from None

    def get_feature(self, feature_id: str) -> Feature:
        try:
            return self._idx.features_by_id[feature_id]
        except KeyError:
            raise NotFoundError(f"Feature not found: {feature_id!r}") from None

    def list_makes(self) -> list[Make]:
        return list(self._idx.makes_by_id.values())

    def list_models(self, make_id: str | None = None) -> list[Model]:
        if make_id is None:
            return list(self._idx.models_by_id.values())
        return list(self._idx.models_by_make.get(make_id, []))

    def list_trims(self, model_id: str | None = None) -> list[Trim]:
        if model_id is None:
            return list(self._idx.trims_by_id.values())
        return list(self._idx.trims_by_model.get(model_id, []))

    def list_features(
        self,
        brand_scope: str | None = None,
        category: str | None = None,
    ) -> list[Feature]:
        out: list[Feature] = []
        for f in self._idx.features_by_id.values():
            if brand_scope is not None and f.brand_scope != brand_scope:
                continue
            if category is not None and f.category != category:
                continue
            out.append(f)
        return out

    def list_features_for_trim(
        self,
        trim_id: str,
        availability: Iterable[str] = ("standard",),
    ) -> list[Feature]:
        if trim_id not in self._idx.trims_by_id:
            raise NotFoundError(f"Trim not found: {trim_id!r}")
        wanted = set(availability)
        out: list[Feature] = []
        seen_feature_ids: set[str] = set()
        for tf in self._idx.features_by_trim.get(trim_id, []):
            if tf.availability not in wanted:
                continue
            if tf.feature_id in seen_feature_ids:
                continue
            feat = self._idx.features_by_id.get(tf.feature_id)
            if feat is None:
                continue
            seen_feature_ids.add(tf.feature_id)
            out.append(feat)
        return out

    def list_trims_with_feature(self, feature_id: str) -> list[Trim]:
        if feature_id not in self._idx.features_by_id:
            raise NotFoundError(f"Feature not found: {feature_id!r}")
        out: list[Trim] = []
        seen_trim_ids: set[str] = set()
        for tf in self._idx.trims_by_feature.get(feature_id, []):
            if tf.availability == "unavailable":
                continue
            if tf.trim_id in seen_trim_ids:
                continue
            trim = self._idx.trims_by_id.get(tf.trim_id)
            if trim is None:
                continue
            seen_trim_ids.add(tf.trim_id)
            out.append(trim)
        return out

    def validate(self) -> ValidationResult:
        trim_features = [
            tf for cells in self._idx.features_by_trim.values() for tf in cells
        ]
        return Validator.check(
            makes=list(self._idx.makes_by_id.values()),
            models=list(self._idx.models_by_id.values()),
            trims=list(self._idx.trims_by_id.values()),
            features=list(self._idx.features_by_id.values()),
            trim_features=trim_features,
        )

File: vehicle-feature-catalog/src/ts/facade.ts
import { Feature, Make, Model, Trim, ValidationResult } from './entities.js';
import { NotFoundError } from './errors.js';
import { CatalogIndexes, buildIndexes } from './indexes.js';
import { YAMLLoader } from './loader.js';
import { validateCatalog } from './validator.js';

export class VehicleFeatureCatalog {
  private readonly idx: CatalogIndexes;

  private constructor(indexes: CatalogIndexes) {
    this.idx = indexes;
  }

  static async load(dataDir: string): Promise<VehicleFeatureCatalog> {
    const raw = await YAMLLoader.read_all(dataDir);
    const idx = buildIndexes(
      raw.makes,
      raw.models,
      raw.trims,
      raw.features,
      raw.trim_features,
    );
    return new VehicleFeatureCatalog(idx);
  }

  get_make(make_id: string): Make {
    const v = this.idx.makes_by_id.get(make_id);
    if (v === undefined) {
      throw new NotFoundError(`Make not found: '${make_id}'`);
    }
    return v;
  }

  get_model(model_id: string): Model {
    const v = this.idx.models_by_id.get(model_id);
    if (v === undefined) {
      throw new NotFoundError(`Model not found: '${model_id}'`);
    }
    return v;
  }

  get_trim(trim_id: string): Trim {
    const v = this.idx.trims_by_id.get(trim_id);
    if (v === undefined) {
      throw new NotFoundError(`Trim not found: '${trim_id}'`);
    }
    return v;
  }

  get_feature(feature_id: string): Feature {
    const v = this.idx.features_by_id.get(feature_id);
    if (v === undefined) {
      throw new NotFoundError(`Feature not found: '${feature_id}'`);
    }
    return v;
  }

  list_makes(): Make[] {
    return Array.from(this.idx.makes_by_id.values());
  }

  list_models(make_id?: string | null): Model[] {
    if (make_id == null) {
      return Array.from(this.idx.models_by_id.values());
    }
    return [...(this.idx.models_by_make.get(make_id) ?? [])];
  }

  list_trims(model_id?: string | null): Trim[] {
    if (model_id == null) {
      return Array.from(this.idx.trims_by_id.values());
    }
    return [...(this.idx.trims_by_model.get(model_id) ?? [])];
  }

  list_features(filters?: {
    brand_scope?: string | null;
    category?: string | null;
  }): Feature[] {
    const brandScope = filters?.brand_scope ?? null;
    const category = filters?.category ?? null;
    const out: Feature[] = [];
    for (const f of this.idx.features_by_id.values()) {
      if (brandScope != null && f.brand_scope !== brandScope) continue;
      if (category != null && f.category !== category) continue;
      out.push(f);
    }
    return out;
  }

  list_features_for_trim(
    trim_id: string,
    availability: ReadonlyArray<'standard' | 'optional'> = ['standard'],
  ): Feature[] {
    if (!this.idx.trims_by_id.has(trim_id)) {
      throw new NotFoundError(`Trim not found: '${trim_id}'`);
    }
    const wanted = new Set<string>(availability);
    const cells = this.idx.features_by_trim.get(trim_id) ?? [];
    const seen = new Set<string>();
    const out: Feature[] = [];
    for (const tf of cells) {
      if (!wanted.has(tf.availability)) continue;
      if (seen.has(tf.feature_id)) continue;
      const feat = this.idx.features_by_id.get(tf.feature_id);
      if (feat === undefined) continue;
      seen.add(tf.feature_id);
      out.push(feat);
    }
    return out;
  }

  list_trims_with_feature(feature_id: string): Trim[] {
    if (!this.idx.features_by_id.has(feature_id)) {
      throw new NotFoundError(`Feature not found: '${feature_id}'`);
    }
    const cells = this.idx.trims_by_feature.get(feature_id) ?? [];
    const seen = new Set<string>();
    const out: Trim[] = [];
    for (const tf of cells) {
      if (tf.availability === 'unavailable') continue;
      if (seen.has(tf.trim_id)) continue;
      const trim = this.idx.trims_by_id.get(tf.trim_id);
      if (trim === undefined) continue;
      seen.add(tf.trim_id);
      out.push(trim);
    }
    return out;
  }

  validate(): ValidationResult {
    const trimFeatures = [...this.idx.features_by_trim.values()].flat();
    return validateCatalog(
      Array.from(this.idx.makes_by_id.values()),
      Array.from(this.idx.models_by_id.values()),
      Array.from(this.idx.trims_by_id.values()),
      Array.from(this.idx.features_by_id.values()),
      trimFeatures,
    );
  }
}
